src/runtime/graph/graph_runtime_factory.cc

```diff
@@ -0,0 +1,196 @@
+      for (int i = 1; i < args.num_args; ++i) {
+        contexts.emplace_back(args[i].operator TVMContext());
+      }
+      *rv = this->DebugRuntimeCreate(module, contexts);
+    });
+  } else if (name == "remove_params") {
+    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
+      auto exec = make_object<GraphRuntimeFactory>();
+      exec->Init(this->GetJson(), {}, this->GetModuleName());
+      exec->Import(this->GetLib());
+      *rv = Module(exec);
+    });
+  } else {
+    return PackedFunc();
+  }
+}
+
+void GraphRuntimeFactory::SaveToBinary(dmlc::Stream* stream) {
+  stream->Write(graph_json_);
+  std::vector<std::string> names;
+  std::vector<DLTensor*> arrays;
+  for (const auto& v : params_) {
+    names.emplace_back(v.first);
+    arrays.emplace_back(const_cast<DLTensor*>(v.second.operator->()));
+  }
+  uint64_t sz = arrays.size();
+  CHECK(sz == names.size());
+  stream->Write(sz);
+  stream->Write(names);
+  for (size_t i = 0; i < sz; ++i) {
+    tvm::runtime::SaveDLTensor(stream, arrays[i]);
+  }
+  stream->Write(module_name_);
+}
+
+Module GraphRuntimeFactory::RuntimeCreate(Module module, const std::vector<TVMContext>& ctxs) {
+  auto factory_module = module.as<GraphRuntimeFactory>();
+  CHECK(factory_module != nullptr);
+  auto exec = make_object<GraphRuntime>();
+  exec->Init(factory_module->GetJson(), factory_module->GetLib(), ctxs);
+  // set params
+  SetParams(exec.get(), factory_module->GetParams());
+  return Module(exec);
+}
+
+Module GraphRuntimeFactory::DebugRuntimeCreate(Module module, const std::vector<TVMContext>& ctxs) {
+  auto factory_module = module.as<GraphRuntimeFactory>();
+  CHECK(factory_module != nullptr);
+  const PackedFunc* pf = tvm::runtime::Registry::Get("tvm.graph_runtime_debug.create");
+  CHECK(pf != nullptr) << "Cannot find function tvm.graph_runtime_debug.create in registry. "
+                          "Do you enable debug graph runtime build?";
+  // Debug runtime create packed function will call GetAllContexs, so we unpack the ctxs.
+  std::vector<int> unpacked_ctxs;
+  for (const auto& ctx : ctxs) {
+    unpacked_ctxs.emplace_back(ctx.device_type);
+    unpacked_ctxs.emplace_back(ctx.device_id);
+  }
+  size_t args_size = unpacked_ctxs.size() + 2;
+  std::vector<TVMValue> values(args_size);
+  std::vector<int> codes(args_size);
+  runtime::TVMArgsSetter setter(values.data(), codes.data());
+  setter(0, factory_module->GetJson());
+  setter(1, factory_module->GetLib());
+  for (size_t i = 0; i < unpacked_ctxs.size(); ++i) {
+    setter(i + 2, unpacked_ctxs[i]);
+  }
+  TVMRetValue rv;
+  pf->CallPacked(TVMArgs(values.data(), codes.data(), args_size), &rv);
+  Module mod = rv.operator Module();
+  // debug graph runtime is one child class of graph runtime.
+  SetParams(const_cast<GraphRuntime*>(mod.as<GraphRuntime>()), factory_module->GetParams());
+  return mod;
+}
+
+Module GraphRuntimeFactory::SelectModule(const std::string& name) {
```

Comment:
The name Select is a bit confusing, let us fold into RuntimeCreate and DebugRuntime create instead, so that we don't need to create a separate factory module